Add IPv6 support to kubenet hostport

aojea opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>
> /kind api-change

/kind bug

> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test

/kind feature

> /kind flake

**What this PR does / why we need it**:

Kubenet uses HostPortManager and HostPortSyncer to implement HostPorts and forward traffic to ports exposed from the pods.

Both of them are only working for IPv4, and since they are used only from the kubenet, that wasn't a problem.

It turns out that the code is vendored by other projects, see bug #80810, and fails if is used with IPv6. In addition, the new dual stack feature allows kubenet to work in dual stack mode.

This PR introduces the following changes:
* Kubenet: adds one handle per IP family for the HostPortManager and HostPortSyncer 
* HostPortManager: operates only with one IP family, it fails if receives portmapping entries with different IP families
* HostPortSyncer: operates only with one IP family, skip portmap entries with different IP families

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #80810

**Special notes for your reviewer**:


**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
Kubenet: added HostPort IPv6 support
HostPortManager: operates only with one IP family, failing if receives portmapping entries with different IP families
HostPortSyncer: operates only with one IP family, skipping portmap entries with different IP families
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.
Please use the following format for linking documentation or pass the
section below:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
```

aojea committed:
Add IPv6 support to kubenet hostport

aojea commented:
/assign @dchen1107

BenTheElder commented on pkg/kubelet/dockershim/network/hostport/hostport_manager.go:
```diff
@@ -82,7 +83,7 @@ func (hm *hostportManager) Add(id string, podPortMapping *PodPortMapping, natInt
 	}
 
-	if podPortMapping.IP.To4() == nil {
+	if podPortMapping.IP.To4() == nil && podPortMapping.IP.To16() == nil {
```
should we use the net utils for this?

aojea replied:
which function do you suggest?

BenTheElder replied:
you're right, we don't seem to have one.

mattjmcnaughton reviewed (commented):
Thanks for your diff @aojea - especially the updated test coverage :)

A couple of questions from me (and I apologize for my lack of familiarity w/ IPv6).

1. What is the overall story of IPV6 support in k8s? Is this part of some broader initiative or are we fixing a one-off failure here. Or is it a bug that host port manager doesn't support IPV6?

2. Confirming this change is only necessary for `dockershim`? From grepping the source code, I'm only seeing references to `hostport_manager` in `dockershim`, so I assume the answer is yes. But just wanted to double check.

3. You mentioned that the master of CNI supports IPV6. Do you know if k8s is actually using the master version? If not, does it make sense for us wait on this change until k8s is upgraded to use a CNI version that supports IPV6?

aojea commented:
> 1. What is the overall story of IPV6 support in k8s? Is this part of some broader initiative or are we fixing a one-off failure here. Or is it a bug that host port manager doesn't support IPV6?

IPv6 was released as alpha in 1.9, and I´m trying to graduate it to beta adding e2e testing https://github.com/kubernetes/enhancements/pull/1139. However, current e2e testing is using containerd only and this was discovered using CRI-O per the issue reported #80810

> 2. Confirming this change is only necessary for `dockershim`? From grepping the source code, I'm only seeing references to `hostport_manager` in `dockershim`, so I assume the answer is yes. But just wanted to double check.

e2e tests are running constantly with containerd as runtime https://testgrid.k8s.io/conformance-kind#kind%20(IPv6),%20master%20(dev), however, don´t know if those will catch this bug.

> 3. You mentioned that the master of CNI supports IPV6. Do you know if k8s is actually using the master version? If not, does it make sense for us wait on this change until k8s is upgraded to use a CNI version that supports IPV6?

CNI has supported IPv6 since 0.6.0 and k8s since 1.9 https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG-1.9.md#ipv6 , my understanding is that if something doesn´t work with IPv6 in k8s now is a bug, right?

BenTheElder commented:
[ipv6 testing specifically on kind => contianerd, other e2e testing has dockershim, ...]

BenTheElder commented:
/priority important-longterm
/assign @thockin

johscheuer commented on pkg/kubelet/dockershim/network/hostport/hostport_syncer.go:
```diff
@@ -256,11 +257,12 @@ func (h *hostportSyncer) SyncHostports(natInterfaceName string, activePodPortMap
 
 		// Create hostport chain to DNAT traffic to final destination
 		// IPTables will maintained the stats for this chain
+		hostPortBinding := net.JoinHostPort(target.podIP, fmt.Sprintf("%d", port.ContainerPort))
```
Same here like above.

aojea commented:
/test pull-kubernetes-integration

BenTheElder commented on pkg/kubelet/dockershim/network/hostport/hostport_syncer.go:
```diff
@@ -122,7 +123,7 @@ func getPodFullName(pod *PodPortMapping) string {
 	podHostportMap := make(map[*PortMapping]targetPod)
 	for _, pm := range activePodPortMappings {
-		if pm.IP.To4() == nil {
+		if pm.IP.To4() == nil && pm.IP.To16() == nil {
```
isn't this the same as just:
```suggestion
		if pm.IP.To16() == nil {
```

aojea replied:
yeah, didn't know that IP.To16() works for IPv4 and IPv6 addresses.
It also applies to https://github.com/kubernetes/kubernetes/pull/80854/files#diff-b0ee819c32e4070b3a2cb6376b72408bR86

aojea replied:
my concern is if this will be clear for another person reading the code, the current `if statement` makes clear that checks that the IP address is neither IPv4 nor IPV6, using only To16() can lead to confusion IMHO

BenTheElder replied:
I disagree, unless they are unclear on what To16() is. However, if you wanted to be most clear I'd wrap this in a method like `isValid(ip)`

aojea replied:
done on both files

danwinship commented:
> > 1. Confirming this change is only necessary for `dockershim`? From grepping the source code, I'm only seeing references to `hostport_manager` in `dockershim`, so I assume the answer is yes. But just wanted to double check.
> 
> e2e tests are running constantly with containerd as runtime https://testgrid.k8s.io/conformance-kind#kind%20(IPv6),%20master%20(dev), however, don´t know if those will catch this bug.

HostPort handling is kind of bizarre. K8s itself only uses this code if you are using docker with the kubenet plugin. IIRC, containerd uses the CNI portmap plugin to do this, so if that's what we're using in our e2e environment then fixes here would not get picked up in our e2e tests. (Although they would *eventually* get tested by OpenShift's e2e tests, which use cri-o, which vendors the k8s HostPort code...)

danwinship commented on pkg/kubelet/dockershim/network/hostport/hostport_manager.go:
```diff
@@ -142,10 +143,11 @@ func (hm *hostportManager) Add(id string, podPortMapping *PodPortMapping, natInt
 			"-m", "comment", "--comment", fmt.Sprintf(`"%s hostport %d"`, podFullName, pm.HostPort),
 			"-m", protocol, "-p", protocol,
-			"-j", "DNAT", fmt.Sprintf("--to-destination=%s:%d", podIP, pm.ContainerPort))
+			"-j", "DNAT", fmt.Sprintf("--to-destination=%s", hostPortBinding))
```
`hostportManager.iptables` / `hostportSyncer.iptables` is specifically either an IPv4-only interface or an IPv6-only interface and maybe we should make sure it's the right one? (kubenet is the only user of this code in-tree, and it passes an IPv4-only interface, but I think it also only supports IPv4 pod IPs so that's ok)

aojea replied:
I' ve added checks to verify that the pod IP and the iptables version are the same

BenTheElder commented:
cc @Random-Liu

aojea commented:
> HostPort handling is kind of bizarre. K8s itself only uses this code if you are using docker with the kubenet plugin. IIRC, containerd uses the CNI portmap plugin to do this, so if that's what we're using in our e2e environment then fixes here would not get picked up in our e2e tests. (Although they would _eventually_ get tested by OpenShift's e2e tests, which use cri-o, which vendors the k8s HostPort code...)

Based on your explanation seems that is more an OpenShift issue than a kubernetes issue.
Should we close the bug and the PR?

danwinship commented:
Well, I think other people are vendoring the kubelet code as well. And when I said the code would eventually get tested in OpenShift, I meant *if you fix it here now*, it will eventually get tested in OpenShift. If the upstream never gets fixed, the downstreams will never get fixed either.

Really the code ought to be moved to another repo like k/utils probably. But while it still exists in k/k, we should fix it here.

aojea commented:
> Well, I think other people are vendoring the kubelet code as well. And when I said the code would eventually get tested in OpenShift, I meant _if you fix it here now_, it will eventually get tested in OpenShift. If the upstream never gets fixed, the downstreams will never get fixed either.
> 
> Really the code ought to be moved to another repo like k/utils probably. But while it still exists in k/k, we should fix it here.

@danwinship sorry, I was understanding just the opposite of your comment :sweat_smile: 

I've addressed the comments, but I have one question about the unit tests, how do you use to do in these situations, are you replicating the same tests for each family?

danwinship commented:
I'm not sure there's much of a standard yet. In a case like this where testing both IPv4 and IPv6 isn't going to make the test runs noticeably slower, it's probably good to test them both.

thockin commented on pkg/kubelet/dockershim/network/hostport/hostport_manager.go:
```diff
@@ -82,11 +83,15 @@ func (hm *hostportManager) Add(id string, podPortMapping *PodPortMapping, natInt
 	}
 
-	if podPortMapping.IP.To4() == nil {
+	if podPortMapping.IP.To16() == nil {
```
comment on the fact that this is safe for valid v4 IPs

aojea replied:
done

thockin commented on pkg/kubelet/dockershim/network/hostport/hostport_manager.go:
```diff
@@ -82,11 +83,15 @@ func (hm *hostportManager) Add(id string, podPortMapping *PodPortMapping, natInt
 	}
 	podIP := podPortMapping.IP.String()
 
+	if utilnet.IsIPv6String(podIP) != hm.iptables.IsIpv6() {
```
with multiple pod IPs, should this be scanning the list looking for any match?  @khenidak

thockin commented on pkg/kubelet/dockershim/network/hostport/hostport_syncer.go:
```diff
@@ -227,6 +229,9 @@ func (h *hostportSyncer) SyncHostports(natInterfaceName string, activePodPortMap
 	for port, target := range hostportPodMap {
 		protocol := strings.ToLower(string(port.Protocol))
 		hostportChain := hostportChainName(port, target.podFullName)
+		if utilnet.IsIPv6String(target.podIP) != h.iptables.IsIpv6() {
```
same comment about multi-IP

aojea replied:
I've changed the approach so kubenet initiates one hostport_manager and hostport_syncer per IP family. If the pod IP doesn't match the IP familiy of the hostport_manager/syncer it simply skips it

aojea replied:
hostport syncer and manager only works with the IP family they have configured
hostport manager fails if the pod IP family is different.
hostport syncer only gets the portMaps from the Pods with the same IP family is configured, skipping the others

aojea commented:
/assign @khenidak

aojea commented:
/retest

thockin commented:
This looks OK, but I admit to being a bit stale on this code.  what I don't see is feature-gating.  We have is a gate `IPv6DualStack` which is supposed to ensure that no new risks are incurred when it is not enabled.

khenidak commented on pkg/kubelet/dockershim/network/hostport/hostport.go:
```diff
@@ -136,7 +136,11 @@ func ensureKubeHostportChains(iptables utiliptables.Interface, natInterfaceName
+			localhost = "::1/128"
+		}
+		args = []string{"-m", "comment", "--comment", "SNAT for localhost access to hostports", "-o", natInterfaceName, "-s", localhost, "-j", "MASQUERADE"}
 		if _, err := iptables.EnsureRule(utiliptables.Append, utiliptables.TableNAT, utiliptables.ChainPostrouting, args...); err != nil {
```
The iptables interface is different `IPv6` or `IPv4` each one calls a different binary. Shouldn't we load the correct iptables interface here? or is this done early on?

aojea replied:
`iptables` is received as a parameter, this function it's called from the hostport_manager or hostport_syncer methods that only handle one IP family

khenidak commented:
@thockin  my understanding is, this is for `IPv6` support, which is for:
1- dualstack
2- Single stack `ipv6` kubenet
3- some users are vendoring these into their code (I am not fully aware who).

aojea commented on pkg/kubelet/dockershim/network/kubenet/kubenet_linux.go:
```diff
@@ -520,6 +531,14 @@ func (plugin *kubenetNetworkPlugin) teardown(namespace string, name string, id k
 		}); err != nil {
 			errList = append(errList, err)
 		}
+		if err = plugin.hostportManagerv6.Remove(id.ID, &hostport.PodPortMapping{
```
I was struggling with this, and I don't have clear the best approach. Let me try to share my thoughts:

With the Add() method we can choose the handler because we know the IP family beforehand.
However, with Remove() we receive a list of `portMappings` that can have IPv4 and IPv6 

https://github.com/kubernetes/kubernetes/blob/6d8dd211111474f277ef8e390a3eeed2465fb9e0/pkg/kubelet/dockershim/docker_service.go#L377

Also, Remove() doesn't take IP addresses on consideration, it finds the iptables chains based on the hash.

https://github.com/kubernetes/kubernetes/blob/a33840e0232aaf2b5c14c1c754d9189d7e841ec1/pkg/kubelet/dockershim/network/hostport/hostport_manager.go#L311

Maybe I should process the list of portMappings and create one list per IP family?

aojea replied:
PTAL

aojea commented:
/test pull-kubernetes-bazel-build

aojea commented:
> @aojea - Can you please let me know if you are working towards getting this merged for 1.16? The code freeze deadline is Thursday 8/29 EoD US Pacific.

I let @khenidak decide since he 's driving the dual-stack wagon :)

khenidak commented:
Just as i said on slack. I think this PR is really IPv6 support in host port, which is not exactly dualstack, yet it is/will be used by dualstack. So I think we need to 
1. rename the PR (and make sure that release notes reflect that)
2. rebase

I am looking now one final look, but it is looking ok.

khenidak commented:
i think this looks ok, will lgtm once rebased.

aojea commented on pkg/kubelet/dockershim/network/kubenet/kubenet_linux.go:
```diff
@@ -459,8 +463,14 @@ func (plugin *kubenetNetworkPlugin) addPortMapping(id kubecontainer.ContainerID,
+			if err := plugin.hostportManagerv6.Add(id.ID, pm, BridgeName); err != nil {
+				return err
+			}
+		} else {
```
these calls are inside a for loop (L458), we can't return, can we?

khenidak commented:
/retest

khenidak commented:
/lgtm

khenidak commented:
/retest

khenidak commented:
/test pull-kubernetes-verify

khenidak commented:
/test pull-kubernetes-integration

aojea commented:
/test pull-kubernetes-e2e-gce

aojea commented:
/test pull-kubernetes-kubemark-e2e-gce-big

khenidak commented:
/lgtm
/approve
/retest

khenidak commented:
/assign @dchen1107 

@dchen1107 can we approve this, host ports support for ipv6 needed by both ipv6 single stack and dualstack.

dcbw commented:
@aojea need to revert the move of the cached pod IPs check; once that's done everything else LGTM and I'll approve/lgtm.

aojea commented on pkg/kubelet/dockershim/network/kubenet/kubenet_linux.go:
```diff
@@ -499,40 +509,64 @@ func (plugin *kubenetNetworkPlugin) SetUpPod(namespace string, name string, id k
-	if !exists || len(iplist) == 0 {
+	if len(iplist) == 0 {
 		klog.V(5).Infof("container %s (%s/%s) does not have recorded. ignoring teardown call", id, name, namespace)
 		return nil
```
I still don't understand this very well, why we have to return here. Do we have to wait for the `delContainerFromNetwork` calls? 
if affirmative, then why we return nil instead of the errors?

aojea commented:
/retest

aojea committed:
Kubenet can't fail fast on teardown

BenTheElder commented:
gentle ping @dcbw @khenidak

hangzhouwanjun commented:
Hi, could you merge support hostport in IPv6 single stack, we want use hostport in IPv6 single stack!
Thank you!

thockin commented:
This LGTM - @dcbw and @khenidak ?

/approve

dims commented:
/milestone v1.17
/lgtm

liggitt commented:
/test pull-kubernetes-conformance-kind-ipv6

liggitt commented:
/test pull-kubernetes-e2e-kind

liggitt commented:
/retest
